File: 3635-earliest-finish-time-for-land-and-water-rides-ii/3635-earliest-finish-time-for-land-and-water-rides-ii.py
class Solution(object):
    def earliestFinishTime(self, landStartTime, landDuration, waterStartTime, waterDuration):
        def calc(s1, d1, s2, d2):
            min_end= float('inf')
            for i in range(len(s1)):
                fin= s1[i]+d1[i]
                if fin<min_end:
                    min_end= fin
            ans= float('inf')
            for i in range(len(s2)):
                fin_t= max(min_end, s2[i])+d2[i]
                if fin_t<ans:
                    ans=fin_t
            return ans
        mioaw=calc(landStartTime, landDuration, waterStartTime, waterDuration)
        meow= calc(waterStartTime, waterDuration, landStartTime, landDuration)
        return min(mioaw,meow)

File: 3635-earliest-finish-time-for-land-and-water-rides-ii/3635-earliest-finish-time-for-land-and-water-rides-ii.java
class Solution {
    public int earliestFinishTime(int[] landStartTime, int[] landDuration, int[] waterStartTime, int[] waterDuration) {
        
        int ans1 = calc(landStartTime, landDuration, waterStartTime, waterDuration);

        int ans2 = calc(waterStartTime, waterDuration, landStartTime, landDuration);

        return Math.min(ans1, ans2);
    }

    private int calc(int[] s1, int[] d1, int[] s2, int[] d2) {

        int minEnd = Integer.MAX_VALUE;

        for (int i = 0; i < s1.length; i++) {

            int fin = s1[i] + d1[i];

            if (fin < minEnd) {
                minEnd = fin;
            }
        }

        int ans = Integer.MAX_VALUE;

        for (int i = 0; i < s2.length; i++) {

            int finT = Math.max(minEnd, s2[i]) + d2[i];

            if (finT < ans) {
                ans = finT;
            }
        }

        return ans;
    }
}